src/replay/consensus.zig

```diff
@@ -0,0 +1,1243 @@
+/// potentially processing a new root, and pushing the vote.
+///
+/// Analogous to [handle_votable_bank](https://github.com/anza-xyz/agave/blob/ccdcdbe9b6ff7dbd583d2101fe57b7cc41a6f863/core/src/replay_stage.rs#L2388)
+fn handleVotableBank(
+    allocator: std.mem.Allocator,
+    ledger_result_writer: *LedgerResultWriter,
+    vote_slot: Slot,
+    vote_hash: Hash,
+    slot_tracker: *SlotTracker,
+    replay_tower: *ReplayTower,
+    progress: *ProgressMap,
+    fork_choice: *ForkChoice,
+) !void {
+    const maybe_new_root = try replay_tower.recordBankVote(
+        allocator,
+        vote_slot,
+        vote_hash,
+    );
+
+    if (maybe_new_root) |new_root| {
+        try checkAndHandleNewRoot(
+            allocator,
+            ledger_result_writer,
+            slot_tracker,
+            progress,
+            fork_choice,
+            new_root,
+        );
+    }
+
+    // TODO update_commitment_cache
+
+    try pushVote(
+        replay_tower,
+    );
+}
+
+/// Pushes a new vote transaction to the network and updates tower state.
+///
+/// - Generates a new vote transaction.
+/// - Updates the tower's last vote blockhash.
+/// - Creates a saved tower state.
+/// - Sends the vote operation to the voting sender.
+///
+/// Analogous to [push_vote](https://github.com/anza-xyz/agave/blob/ccdcdbe9b6ff7dbd583d2101fe57b7cc41a6f863/core/src/replay_stage.rs#L2775)
+fn pushVote(
+    replay_tower: *ReplayTower,
+) !void {
+
+    // TODO Transaction generation to be implemented.
+    // Currently hardcoding to non voting transaction.
+    const vote_tx_result: GenerateVoteTxResult = .non_voting;
+
+    switch (vote_tx_result) {
+        .tx => |vote_tx| {
+            _ = &vote_tx;
+            // TODO to be implemented
+        },
+        .non_voting => {
+            replay_tower.markLastVoteTxBlockhashNonVoting();
+        },
+        else => {
+            // Do nothing
+        },
+    }
+}
+
+/// Processes a new root slot by updating various system components to reflect the new root.
+///
+/// - Validates the new root exists and has a hash.
+/// - Gets all slots rooted at the new root.
+/// - Updates ledger state with the new roots.
+/// - Cleans up progress map for non-existent slots.
+/// - Updates fork choice with the new root.
+///
+/// Analogous to [check_and_handle_new_root](https://github.com/anza-xyz/agave/blob/ccdcdbe9b6ff7dbd583d2101fe57b7cc41a6f863/core/src/replay_stage.rs#L4002)
+fn checkAndHandleNewRoot(
+    allocator: std.mem.Allocator,
+    ledger_result_writer: *LedgerResultWriter,
+    slot_tracker: *SlotTracker,
+    progress: *ProgressMap,
+    fork_choice: *ForkChoice,
+    new_root: Slot,
+) !void {
+    // get the root bank before squash.
+    if (slot_tracker.slots.count() == 0) return error.EmptySlotTracker;
+    var root_tracker = slot_tracker.slots.get(new_root) orelse return error.MissingSlot;
+    const maybe_root_hash, var hash_lg = root_tracker.state.hash.readWithLock();
```

Comment:
Same comment as on the other instance of `slot_tracker.slots.get(slot)`.
Also, it should be `const root_tracker`, no reason for it to be var.

Reply:
[0245b07](https://github.com/Syndica/sig/pull/762/commits/0245b075510c2b16402df21b442516033de01709)